### PR TITLE
[mypyc] Optimize str.encode with specializations for common used encodings

### PR DESCRIPTION
**Pull request by svalentin**
Tested with:
```
import time
start = time.time()
for i in range(20000000):
    "test".encode('utf-8')
print(time.time() - start)
```

With PR applied and running mypyc, `python3 -c "import test"` runs in:
0.5383486747741699
0.5224344730377197
0.555696964263916

Without PR applied:
0.7315819263458252
0.7105758190155029
0.7471706867218018

Similar times observed for "ascii"

### CODE REVIEW

File: mypyc/test-data/irbuild-str.test [hunk 1]
```diff
@@ -293,20 +293,29 @@ L0:
 def f(s: str) -> None:
     s.encode()
     s.encode('utf-8')
+    s.encode('utf-8', 'strict')
+    s.encode('utf-8', errors='strict')
+    s.encode('utf-8', 'backslashreplace')
+    s.encode(encoding='ascii')
     s.encode('ascii', 'backslashreplace')
```
Comment (JukkaL) [inline mypyc/test-data/irbuild-str.test:300]: Also test cases where the specialization shouldn't be applied. Examples: `s.encode(x)`, `s.encode('a', x)`, `s.encode('utf8', errors=x)` and `s.encode(errors=x)` where `x` is not a literal.

Test cases where we have two keyword args: `s.encode(encoding=..., errors=...)` and `s.encode(errors=..., encoding=...)`.

**Review by JukkaL — COMMENTED** [review]
I think some cases aren't covered by the logic. Suggested some test cases that should help.

File: mypyc/test-data/irbuild-str.test (hunk 1 above)
Comment (svalentin) [reply mypyc/test-data/irbuild-str.test:301]: I've updated the logic to work out the args better and added more tests. Please take another look!

**Review by JukkaL — APPROVED** [review]
Looks good, thanks!

**Pull request merged**